src/util/vrf.rs

```diff
@@ -192,214 +380,202 @@ impl ECVRF_Proof {
+    }
 
-    let U_point = &s_reduced * &RISTRETTO_BASEPOINT_POINT - proof.c() * Y_ristretto_point;
-    let V_point = &s_reduced * &H_point - proof.c() * proof.Gamma();
+    /// Given a public key, verify that the private key owner that generate the ECVRF proof did so on the given message.
+    /// Return Ok(true) if so
+    /// Return Ok(false) if not
+    /// Return Err(Error) if the public key is invalid, or we are unable to do one of the
+    /// necessary internal data conversions.
+    pub fn verify(Y_point: &VRFPublicKey, proof: &VRFProof, alpha: &Vec<u8>) -> Result<bool, Error> {
+        let H_point = VRF::hash_to_curve(Y_point, alpha);
+        let Y_ristretto_point = VRF::VRFPublicKey_to_RistrettoPoint(Y_point)?;
+        let s_reduced = proof.s().reduce();
 
-    let c_prime_hashbuf = ECVRF_hash_points(&H_point, proof.Gamma(), &U_point, &V_point);
-    let c_prime = ECVRF_ed25519_scalar_from_hash128(&c_prime_hashbuf);
+        let U_point = &s_reduced * &RISTRETTO_BASEPOINT_POINT - proof.c() * Y_ristretto_point;
+        let V_point = &s_reduced * &H_point - proof.c() * proof.Gamma();
 
-    // NOTE: this leverages constant-time comparison inherited from the Scalar impl
-    Ok(c_prime == *(proof.c()))
-}
+        let c_prime_hashbuf = VRF::hash_points(&H_point, proof.Gamma(), &U_point, &V_point);
+        let c_prime = VRF::ed25519_scalar_from_hash128(&c_prime_hashbuf);
 
-/// Verify that a given byte string is a well-formed EdDSA public key (i.e. it's a compressed
-/// Edwards point that is valid).
-pub fn ECVRF_check_public_key(pubkey_bytes: &Vec<u8>) -> Option<ed25519_PublicKey> {
-    match pubkey_bytes.len() {
-        32 => {
-            let mut pubkey_slice = [0; 32];
-            pubkey_slice.copy_from_slice(&pubkey_bytes[0..32]);
-
-            let checked_pubkey = CompressedEdwardsY(pubkey_slice);
-            let full_checked_pubkey = checked_pubkey.decompress();
-            if full_checked_pubkey.is_none() {
-                // invalid 
-                return None;
-            }
-
-            let key_res = ed25519_PublicKey::from_bytes(&pubkey_slice);
-            match key_res {
-                Ok(key) => Some(key),
-                Err(_e) => None
-            }
-        },
-        _ => None
+        // NOTE: this leverages constant-time comparison inherited from the Scalar impl
+        Ok(c_prime == *(proof.c()))
     }
-}
 
-/// Helper method to turn a public key into a hex string 
-pub fn ECVRF_public_key_to_hex(pubkey: &ed25519_PublicKey) -> String {
-    to_hex(pubkey.as_bytes())
+    /// Verify that a given byte string is a well-formed EdDSA public key (i.e. it's a compressed
+    /// Edwards point that is valid).
+    pub fn check_public_key(pubkey_bytes: &Vec<u8>) -> Option<VRFPublicKey> {
+        match pubkey_bytes.len() {
+            32 => {
+                VRFPublicKey::from_bytes(&pubkey_bytes[..])
+            },
+            _ => None
+        }
+    }
 }
 
 #[cfg(test)]
```

Comment:
Should this VRF implementation be able to use the test vectors supplied in https://tools.ietf.org/id/draft-irtf-cfrg-vrf-04.html#rfc.appendix.A?

Reply:
If any apply, it would be those in A.4 -- specifically, the Ed25519-SHA512-Elligator2 test vectors.  However, I'm not sure they apply at all, since our VRF doesn't use the Elligator2 map (but uses the Elligator map on the Ristretto subgroup, defined here: https://ristretto.group/details/elligator.html).  In fact, our hash-to-curve routine (line 398) does not currently use the RFC-recommended hash-to-curve routine for Ed25519-SHA512-Elligator2 (defined here: https://tools.ietf.org/id/draft-irtf-cfrg-vrf-02.html#suites) to avoid confusion.

Reply:
According to curve25519_dalek -- Elligator2 is the map used (see https://doc.dalek.rs/curve25519_dalek/ristretto/struct.RistrettoPoint.html#method.hash_from_bytes). Is there a reason to implement our own hash-to-curve routine? That `hash_from_bytes` function implements hash-to-curve for a given digest.

Reply:
I stand corrected!  I thought they were different Elligator maps.

> Is there a reason to implement our own hash-to-curve routine?

By this, I mean we use a different suite version number (which gets hashed with the public key and alpha).  I'll try it out with the suite version recommended by the RFC.  If we can use the RFC test vectors, we definitely should :+1:

Reply:
Fun story -- the Ristretto algorithm for converting a hash to a point is _not_ the same process for converting a hash to an Ed25519 point.  The Ristretto algorithm applies the Elligator map _twice_ -- once on the upper 32 bytes of the hash, and once on the lower 32 bytes of the hash -- and adds the result.  This is not described by the RFC.

Interestingly, Libra's VRF _does_ follow the RFC, and even uses the same libraries we do: https://github.com/libra/libra/blob/master/crypto/crypto/src/vrf/ecvrf.rs

Reply:
If RFC compliance is the goal here, an alternative is to use a different hash-to-curve algorithm like the try-and-increment algorithm here: https://tools.ietf.org/id/draft-irtf-cfrg-vrf-02.html#ecvrfH2C1.  It's not a constant-time operation, but that's not necessary for us because the alpha parameter (the VRF seed) is public, not secret.

Coincidentally, this is the hash-to-curve algorithm Libra uses in its VRF (not Elligator2).

Reply:
Okay, I re-implemented the VRF to be RFC-compliant.  The ECVRF-ED25519-SHA512-TAI test vectors all pass (we're now using the try-and-increment hash-to-point scheme instead of the elligator map).  Also, all operations are now on full-bore Ed25519 instead of Ristretto points.

Reply:
Awesome!

Reply:
Addressed in a2320797